### PR TITLE
chore(swap): Estimate paid egress fee when egress event is processed

### PR DESCRIPTION
**Pull request by niklasnatter**

### CODE REVIEW

File: packages/swap/src/utils/function.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ export const memoize = <T extends AnyFunction>(fn: T, ttl?: number): T => {
   return ((...args) => {
     if (
       !initialized ||
+      args.some(Boolean) || // TODO: support memoizing result for functions with arguments
```
Comment (niklasnatter) [inline packages/swap/src/utils/function.ts:15]: maybe @acdibble is interested in fixing this

File: packages/sdk/src/swap/sdk.ts [hunk 2]
```diff
@@ -82,7 +82,9 @@ export class SwapSDK {
   }
 
   private async getStateChainEnvironment(): Promise<Environment> {
-    this.stateChainEnvironment ??= await getEnvironment(this.rpcConfig);
+    this.stateChainEnvironment ??= await getEnvironment(this.rpcConfig, [
+      undefined,
```
Comment (niklasnatter) [inline packages/sdk/src/swap/sdk.ts:86]: i havent figured out how to type this nicer, but i think its not too bad

File: packages/shared/src/rpc/index.ts [hunk 3]
```diff
@@ -170,7 +169,11 @@ const environment = z.object({
   pools: poolsEnvironment,
 });
 
-export const getEnvironment = createRequest('cf_environment', environment);
+export const getEnvironment = createRequest(
+  'cf_environment',
+  environment,
+  z.tuple([z.string().optional()]), // [atBlockHash]
```
Comment (niklasnatter) [inline packages/shared/src/rpc/index.ts:175]: would be nice to give this some name but i think we cant?

File: packages/swap/src/utils/rpc.ts [hunk 4]
```diff
@@ -13,30 +13,36 @@ type Result = { success: true } | { success: false; reason: string };
...
 export const getEgressFee = async (
   asset: UncheckedAssetAndChain,
+  blockHash?: string,
 ): Promise<bigint> => {
-  const env = await cachedGetEnvironment({
-    rpcUrl: process.env.RPC_NODE_HTTP_URL as string,
-  });
+  const env = await cachedGetEnvironment(
+    { rpcUrl: process.env.RPC_NODE_HTTP_URL as string },
+    [blockHash],
```
Comment (niklasnatter) [inline packages/swap/src/utils/rpc.ts:44]: im afraid this will break indexing swaps before the upgrade because the returned value wont be valid 😵‍💫

**Review by acdibble — APPROVED** [review]

**Review by jerryafr — APPROVED** [review]

**Pull request merged**